jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -89,6 +90,22 @@ public class InternalDowncallHandler {
 	private MemoryLayout realReturnLayout;
 	private MethodHandle boundMH;
 
+	/*[IF JAVA_SPEC_VERSION >= 22]*/
+	/* The identifier denotes the passed-in argument is a heap segment which helps
+	 * extract the address from the heap argument's base object in native.
+	 */
+	private static final long DOWNCALL_HEAP_ARGUMENT_ID = 0x1;
+
+	/* The hash map holds a record of the base object array, the offset array plus
+	 * the current heap argument index for the current thread in multithreading.
+	 * The heap base array stores the base object (which contains the heap address
+	 * as defined in OpenJDK) of the heap arguments to extract the heap address
+	 * from the base object in native.
+	 */
+	private record HeapArgInfo(Object[] bases, long[] offsets, int curArgIdx) {}
+	private final ConcurrentHashMap<Thread, HeapArgInfo> threadHeapArgInfoMap;
```

Comment:
A `record` seems inappropriate here give we need to update `curArgIdx`, and `ThreadLocal` seems preferable to `Map<Thread, HeapArgInfo>`. Suggest
```java
	private static final class HeapArgInfo {
		final Object[] bases;
		final long[] offsets;
		int index;

		HeapArgInfo(int size) {
			bases = new Object[size];
			offsets = new long[size];
			index = 0;
		}

		void append(Object base, long offset) {
			int i = index;
			bases[i] = base;
			offsets[i] = offset;
			index = i + 1;
		}
	}

	private final ThreadLocal<HeapArgInfo> heapArgInfo;
```

Reply:
Agreed and will update the code with `ThreadLocal` against the suggestion above.

Reply:
Updated the code with `ThreadLocal` and verified in tests.